modules/platforms/dotnet/Apache.Ignite/Internal/Compute/Executor/JobLoadContextCache.cs

```diff
@@ -0,0 +1,241 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Apache.Ignite.Internal.Compute.Executor;
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+/// <summary>
+/// Cache for job load contexts.
+/// <para />
+/// Every job load context is associated with a set of deployment unit paths. For example, [unit1, unit2] and [unit1, unit3] sets
+/// represent a different set of assemblies, so they will have different job load contexts. Unit order matters too.
+/// <para />
+/// Loading assemblies and resolving types can take from ~100us to several seconds, depending on the number of assemblies,
+/// type initializers (static constructors), and other factors.
+/// <para />
+/// If a deployment unit U is undeployed, all job load contexts with U in their set will be removed from the cache immediately.
+/// </summary>
+internal sealed class JobLoadContextCache : IDisposable
+{
+    private static readonly long TicksPerMs = Stopwatch.Frequency / 1000L;
+
+    private readonly int _ttlMs;
+
+    /** Main cache of job load contexts. */
+    private readonly Dictionary<DeploymentUnitPaths, (JobLoadContext Ctx, long Ts)> _jobLoadContextCache = new();
+
+    /** Additional map to quickly find all job load contexts that use a given deployment unit path. */
+    private readonly Dictionary<string, List<DeploymentUnitPaths>> _deploymentUnitSets = new();
+
+    private readonly SemaphoreSlim _cacheLock = new(1);
+
+    private readonly CancellationTokenSource _cts = new();
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="JobLoadContextCache"/> class.
+    /// </summary>
+    /// <param name="ttlMs">Cache entry time-to-live in milliseconds.</param>
+    /// <param name="cacheCleanupIntervalMs">Cache cleanup interval.</param>
+    internal JobLoadContextCache(int ttlMs = 30_000, int cacheCleanupIntervalMs = 5_000)
+    {
+        _ttlMs = ttlMs;
+
+        _ = StartCacheCleanupAsync(cacheCleanupIntervalMs);
```

Comment:
JFYI - AFAIK .NET Memory cache is using lazy expiration, i.e. it will trigger expirations on access. Not sure that this will be a better approach here but maybe worth discussion.